### PR TITLE
Newsletter implement crm memebership

### PR DESCRIPTION
**Pull request by ComvationAG**

### CODE REVIEW

**Comment by ComvationAG** [comment]
MySQL Querys:
ALTER TABLE `contrexx_module_newsletter_tmp_sending`
  MODIFY  `type` enum('access','newsletter','core','crm') NOT NULL DEFAULT 'newsletter';

ALTER TABLE `contrexx_module_newsletter_rel_crm_membership_newsletter`
  MODIFY  `type` enum('associate', 'include', 'exclude') NOT NULL;

File: modules/Crm/Model/Entity/CrmContact.class.php [hunk 1]
```diff
@@ -93,17 +93,24 @@ function load($id = null)
         if (!empty($this->id)) {
             $query = "SELECT c.id, c.customer_id, c.customer_type,
                              c.customer_name, c.customer_addedby,
-                             c.customer_currency, c.contact_amount,
-                             c.contact_familyname, c.contact_title,
+                             c.customer_currency, c.contact_amount, c.contact_familyname, c.contact_title,
                              c.contact_role, c.contact_customer, c.contact_language,c.company_size,
-                             c.notes, c.contact_type,c.user_account,c.updated_date,c.added_date,
```
Comment (daeppen) [inline modules/Crm/Model/Entity/CrmContact.class.php]: please explain the removal of _c.updated_date_

File: modules/Crm/Model/Entity/CrmContact.class.php [hunk 2]
```diff
@@ -341,7 +355,9 @@ function clean()
         $this->family_name      = '';
         $this->contact_title    = '';
         $this->contact_role     = '';
+        $this->contact_title    = '';
```
Comment (daeppen) [inline modules/Crm/Model/Entity/CrmContact.class.php]: drop redundant code

File: modules/Newsletter/Controller/NewsletterLib.class.php [hunk 3]
```diff
@@ -143,14 +144,15 @@ static public function getLists($excludeHiddenLists=false, $extendedInfo=false)
...
+     * For crm email addresses this will be the system default language by now
      * @param string $email E-mail address of the user
      * @param string $type User type (see constants)
      * @return integer Language ID
      */
     public function getUsersPreferredLanguageId($email, $type) {
         global $objDatabase;
 
-        $userLanguage = \FWLanguage::getDefaultLangId();
+        $userLanguage = \FWLanguage::getDefaultLangId(); // used also for crm
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterLib.class.php]: please fetch language id of crm contact

File: modules/Newsletter/Controller/NewsletterManager.class.php [hunk 4]
```diff
@@ -2825,13 +2856,41 @@ protected function getMailRecipientQuery($mailId, $distinctByType = true) {
                     // 3. access users of one of the selected user groups
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterManager.class.php]: please update the comment above

File: modules/Newsletter/Controller/NewsletterManager.class.php [hunk 5]
```diff
@@ -2816,6 +2837,16 @@ protected function getMailRecipientQuery($mailId, $distinctByType = true) {
                                       FROM `%1$smodule_crm_customer_membership` AS `acrm_membership_exclude` 
                                      WHERE `acrm_membership_exclude`.`membership_id` IN ('.join(',', $crmMembershipFilter['exclude']).'))');
 
+        // select recipients based on selected crm memberships
+        $crmMembershipQuery = '';
+        if($crmMembershipFilter['associate']){
+            $crmMembershipQuery = 'UNION DISTINCT SELECT DISTINCT `crm`.`email`' .($distinctByType ? ', "'.self::USER_TYPE_CRM.'" AS `type`' : '').'
+                                        FROM `' . DBPREFIX . 'module_crm_contacts` AS `contact` 
+                                        LEFT JOIN `' . DBPREFIX . 'module_crm_customer_contact_emails` AS `crm`
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterManager.class.php]: should this not be an inner join?

File: modules/Newsletter/Controller/NewsletterManager.class.php [hunk 6]
```diff
@@ -2816,6 +2837,16 @@ protected function getMailRecipientQuery($mailId, $distinctByType = true) {
                                       FROM `%1$smodule_crm_customer_membership` AS `acrm_membership_exclude` 
                                      WHERE `acrm_membership_exclude`.`membership_id` IN ('.join(',', $crmMembershipFilter['exclude']).'))');
 
+        // select recipients based on selected crm memberships
+        $crmMembershipQuery = '';
+        if($crmMembershipFilter['associate']){
+            $crmMembershipQuery = 'UNION DISTINCT SELECT DISTINCT `crm`.`email`' .($distinctByType ? ', "'.self::USER_TYPE_CRM.'" AS `type`' : '').'
+                                        FROM `' . DBPREFIX . 'module_crm_contacts` AS `contact` 
+                                        LEFT JOIN `' . DBPREFIX . 'module_crm_customer_contact_emails` AS `crm` 
+                                        ON `crm`.`contact_id` = `contact`.`id` ' .
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterManager.class.php]: should we not filter by _crm.is_primary_?

File: modules/Newsletter/Controller/NewsletterManager.class.php [hunk 7]
```diff
@@ -3788,6 +3875,34 @@ private function getNewsletterUserData($id, $type)
...
+
+                $arrUserData['sex'] = '';
+                if($crmUser->__get('contact_gender') == 1){
+                    $arrUserData['sex'] = 'f';
+                } else if($crmUser->__get('contact_gender') == 2){
+                    $arrUserData['sex'] = 'm';
+                }
+                // crm dos not support the following fields:
+                // salutation, birthday, industry_sector, country
+                $arrUserData['email']           = $crmUser->__get('email');
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterManager.class.php]: please access the properties directly. i.e.: _$crmUser->email_

File: modules/Newsletter/Controller/NewsletterManager.class.php [hunk 8]
```diff
@@ -2969,16 +3028,23 @@ function _sendMail()
      * Get the emails from the tmp sending page
      *
      * @author      Stefan Heinemann <sh@adfinis.com>
-     * @return
+     * @return      DBIterator
      */
     protected function getTmpSending($id, $amount) {
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterManager.class.php]: the associated recipient id should be stored in the database table contrexx_module_newsletter_tmp_sending (in _setTmpSending()_) instead of being fetch in this method

File: modules/Newsletter/Controller/NewsletterManager.class.php [hunk 9]
```diff
@@ -3740,7 +3827,7 @@ private function getNewsletterUserData($id, $type)
 
         if (!$id) return $arrUserData;
 
-        switch ($type) {
+        switch (strtolower($type)) {
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterManager.class.php]: why is this required?

File: modules/Newsletter/Controller/NewsletterManager.class.php [hunk 10]
```diff
@@ -2825,13 +2856,41 @@ protected function getMailRecipientQuery($mailId, $distinctByType = true) {
...
+                                     WHERE `membership`.`membership_id` IN ('.join(',', $crmMembershipFilter['exclude']).'))') .
+                    ' AND `contact`.`customer_type` = \'1\'
+                      AND `crm`.`is_primary` = \'1\')' .
+                        (
+                            $checkMailType
+                                ? ' OR `s`.`type` != \'crm\')'
+                                : ')'
+                        )
+                      ;
+    }
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterManager.class.php]: there must be an empty line between two methods

File: modules/Newsletter/Controller/NewsletterManager.class.php [hunk 11]
```diff
@@ -3839,7 +3954,10 @@ public function GetUnsubscribeURL($code, $email, $type = self::USER_TYPE_NEWSLET
     {
         global $_ARRAYLANG, $_CONFIG;
 
-        if ($type == self::USER_TYPE_CORE) {
+        if (
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterManager.class.php]: analogous behaviour is missing in _GetProfileURL()_

File: modules/Newsletter/Controller/NewsletterManager.class.php [hunk 12]
```diff
@@ -2825,13 +2856,41 @@ protected function getMailRecipientQuery($mailId, $distinctByType = true) {
...
+                    (!$crmMembershipFilter['exclude'] ? '' :
+                    ' AND `contact`.`id` NOT IN (
+                                    SELECT `membership`.`contact_id`
+                                      FROM `' . DBPREFIX . 'module_crm_customer_membership` AS `membership` 
+                                     WHERE `membership`.`membership_id` IN ('.join(',', $crmMembershipFilter['exclude']).'))') .
+                    ' AND `contact`.`customer_type` = \'1\'
+                      AND `crm`.`is_primary` = \'1\')' .
+                        (
+                            $checkMailType
+                                ? ' OR `s`.`type` != \'crm\')'
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterManager.class.php]: use constant instead of 'crm'

File: modules/Newsletter/Controller/NewsletterLib.class.php (hunk 3 above)
Comment (ComvationAG) [reply modules/Newsletter/Controller/NewsletterLib.class.php]: The language is only used for unsubscribeURL and profileURL. 
The crm user has neither of them.

So we do not need to fetch the language from crm. This will only slow down the system at the moment.

**Comment by ComvationAG** [comment]
@daeppen 
Thanks for reviewing. I have checked the open issues and the should now all be fixed.
So you can review this again.

**Comment by daeppen** [comment]
- [ ] table `contrexx_module_newsletter_email_link_feedback` does not yet support type `NewsletterLib::USER_TYPE_CRM`

**Comment by daeppen** [comment]
- [ ] please explain why you do not check for the existance of the identified crm contact in Newsletter::trackLink()

File: modules/Newsletter/Controller/Newsletter.class.php [hunk 13]
```diff
@@ -1159,6 +1159,32 @@ public static function displayInBrowser()
...
+            $company                   = $crmUser->linkedCompany;
+            $title                     = $crmUser->contact_title;
+            $position                  = $crmUser->contact_role;
+            $zip                       = $crmUser->zip;
+            $city                      = $crmUser->city;
+            $website                   = $crmUser->url;
+            $phoneOffice               = $crmUser->phone;
+            $phonePrivate              = $crmUser->phone;
+            $phoneMobile               = $crmUser->phone;
+            $fax                       = $crmUser->phone;
```
Comment (daeppen) [inline modules/Newsletter/Controller/Newsletter.class.php]: - [ ] setting `$crmUser->phone` for all phone types is wrong. Proposal: add `$crmUser->phoneWork`, `$crmUser->phoneHome`, `$crmUser->phoneMobile` and `$crmUser->phoneFax`

File: modules/Newsletter/Controller/Newsletter.class.php [hunk 14]
```diff
@@ -1239,13 +1265,16 @@ public static function displayInBrowser()
         // prepare links in content for tracking
         if (is_object($objUser) && $objUser->getId()) {
             $userId = $objUser->getId();
-            $realUser = true;
+            $userType = self::USER_TYPE_ACCESS;
+        } elseif (!empty($crmId)) {
+            $userId = $crmId;
+            $userType = self::USER_TYPE_CRM;
```
Comment (daeppen) [inline modules/Newsletter/Controller/Newsletter.class.php]: - [ ] please explain why you do not check for the existance of the identified crm contact here

File: modules/Newsletter/Controller/NewsletterManager.class.php [hunk 15]
```diff
@@ -3750,6 +3852,43 @@ private function getNewsletterUserData($id, $type)
...
+
+                $arrUserData['sex'] = '';
+                if($crmUser->contact_gender == 1){
+                    $arrUserData['sex'] = 'f';
+                } else if($crmUser->contact_gender == 2){
+                    $arrUserData['sex'] = 'm';
+                }
+
+                $profileAttribute = new \User_Profile_Attribute();
+                $salutations = $profileAttribute->getCoreAttributeTitle();
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterManager.class.php]: - [ ] migrate non-working code

File: modules/Newsletter/Controller/NewsletterManager.class.php [hunk 16]
```diff
@@ -3750,6 +3852,43 @@ private function getNewsletterUserData($id, $type)
...
+                $arrUserData['company']         = $crmUser->linkedCompany;
+                $arrUserData['title']           = $crmUser->contact_title;
+                $arrUserData['position']        = $crmUser->contact_role;
+                $arrUserData['zip']             = $crmUser->zip;
+                $arrUserData['city']            = $crmUser->city;
+                $arrUserData['website']         = $crmUser->url;
+                $arrUserData['phone_office']    = $crmUser->phone;
+                $arrUserData['phone_private']   = $crmUser->phone;
+                $arrUserData['phone_mobile']    = $crmUser->phone;
+                $arrUserData['fax']             = $crmUser->phone;
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterManager.class.php]: - [ ] see analogous comment in [modules/Newsletter/Controller/Newsletter.class.php on line 1187](https://github.com/Cloudrexx/cloudrexx/pull/77/files#diff-3e607f4844cff12acd431014e9cbb7e3R1187)

File: modules/Newsletter/Controller/NewsletterLib.class.php [hunk 17]
```diff
@@ -185,6 +187,58 @@ public function getUsersPreferredLanguageId($email, $type) {
...
+     * Returns the Language ID for a newsletter recipient
+     *
+     * If the recipients's preferred language can not be found, the default
+     * language ID is returned.
+     *
+     * @param integer $id   id of the recipient or the linked access/crm user
+     * @param string  $type User type (see constants)
+     * @return integer Language ID
+     */
+    public function getRecipientLocaleIdByRecipientId($id, $type) {
```
Comment (daeppen) [inline modules/Newsletter/Controller/NewsletterLib.class.php]: - [ ] what is this method used for?

File: modules/Newsletter/Controller/Newsletter.class.php (hunk 14 above)
Comment (ComvationAG) [reply modules/Newsletter/Controller/Newsletter.class.php]: @daeppen There are three different types: access user, crm contacts and newsletter recipients.
The first if in this block catches all access users. This else if catches all crm contacts and the else catches the default newsletter recipients.

So $crmId only exists if we really have a user of type crm. Otherwise, it will be empty.

File: modules/Newsletter/Controller/NewsletterLib.class.php (hunk 17 above)
Comment (ComvationAG) [reply modules/Newsletter/Controller/NewsletterLib.class.php]: @daeppen This method is unused and therefore now dropped

File: modules/Newsletter/Controller/Newsletter.class.php (hunk 14 above)
Comment (daeppen) [reply modules/Newsletter/Controller/Newsletter.class.php]: > So $crmId only exists if we really have a user of type crm. Otherwise, it will be empty.

I don't agree. `$crmId` is set if `$_GET['cId']` has been set. There is no proof that a crm contact by $crmId actually exists. You do have to verify that `$crmUser->load($crmId);` return `true`
